python/hail/methods/impex.py

```diff
@@ -953,6 +953,140 @@ def import_table(paths, key=[], min_partitions=None, impute=False, no_header=Fal
+        CTCTTCTCCCCC kidney  2.5 0    0   0   1
+        CTATATAGCTGC brain   7.0 0    0   3   0
+
+    The field ``Height`` contains floating-point numbers and the field ``Age``
+    contains integers.
+
+    To import this matrix:
+
+    >>> matrix1 = hl.import_matrix_table('data/matrix1.tsv',
+    ...                              row_fields={'Barcode': hl.TString(), 'Tissue': hl.TString(), 'Days':hl.TFloat32()},
+    ...                              key='Barcode')
+
+    If the header information is missing for the row fields, like in the following:
+
+    .. code-block:: text
+
+        $ cat data/matrix2.tsv
+        GENE1  GENE2    GENE3   GENE4
+        TTAGCCAGTCCA brain   1.0 0    0   1   0
+        ATCACTTTTACA kidney  5.5 3    0   2   0
+        CTCTTCTCCCCC kidney  2.5 0    0   0   1
+        CTATATAGCTGC brain   7.0 0    0   3   0
+
+    The row fields get imported as `f0`, `f1`, and `f2`, so we need to do:
+
+    >>> matrix2 = hl.import_matrix_table('data/matrix2.tsv',
+    ...                              row_fields={'f0': hl.TString(), 'f1': hl.TString(), 'f2':hl.TFloat32()},
+    ...                              key='f0')
+    >>> matrix2.rename({'f0': 'Barcode', 'f1': 'Tissue', 'f2': 'Days'})
+
+    Sometimes, the header and row information is missing completely:
+
+    .. code-block:: text
+
+        $ cat data/matrix3.tsv
+        0    0   1   0
+        3    0   2   0
+        0    0   0   1
+        0    0   3   0
+
+    >>> matrix3 = hl.import_matrix_table('data/matrix3.tsv', no_header=True)
+
+    In this case, the file has no row fields, so we use the default
+    row index as a key for the imported matrix table.
+
+    Notes
+    -----
+    All columns to be imported as row fields must be at the start of the row.
+
+    Unlike import_table, no type imputation is done so types must be specified
+    for all columns that should be imported as row fields.
+
+    The header information for row fields is allowed to be missing, if the
+    column IDs are present, but the header must then consist only of tab-delimited
+    column IDs (no row field names).
+
+    If row field names are missing (or the no_header flag is ``True``), fields
+    are imported in column-order as `f0`, `f1`, ... (0-indexed).
+
+    If the no_header flag is ``True``, column IDs will be imported as `col0`,
+    `col1`, ... (also 0-indexed).
+
+    Parameters
+    ----------
+    paths: :obj:`str` or :obj:`list` of :obj:`str`
+        Files to import.
+    row_fields: :obj:`dict` of :obj:`str` to :class:`.Type`
+        Columns to take as row fields in the MatrixTable. They must be located
+        before all entry columns.
+    key: :obj:`str` or :obj:`list` of :obj:`str`
+        Key fields(s). If empty, creates an index `row_id` to use as key.
+    entry_type: :class:`.Type`
+        Type of entries in matrix table. Must be one of: TInt32, TInt64, TFloat32,
+        TFloat64, or TString. Default: TInt64.
```

Comment:
I tried to do this before and it complained because the hail context didn't exist. I could maybe try to do this again now that @tpoterba got rid of it in python.

Reply:
this will work now. Also, use `tint64` instead of `TInt64()`.

Reply:
Is this also the case for the other types?

Reply:
types require creation of hail context only when they're used to call into java or for a couple of other things like `pretty`. I made most things lazy in a PR last week

Reply:
cool!